pkg/service/worker_service_test.go

```diff
@@ -0,0 +1,498 @@
+package service_test
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/MinhNghiaD/jobworker/api/client"
+	"github.com/MinhNghiaD/jobworker/api/worker/proto"
+	"github.com/MinhNghiaD/jobworker/pkg/service"
+	"github.com/sirupsen/logrus"
+)
+
+// NOTE we won't use context timeout for grpc testing because the respond time in test is not fast enough
+
+func TestStartJobs(t *testing.T) {
+	server, err := service.NewServer(7777)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	go server.Serve()
+
+	client, err := client.New("127.0.0.1:7777")
+	if err != nil {
+		t.Fatalf("Fail to init client %s", err)
+	}
+
+	defer t.Cleanup(func() {
+		server.Close()
+		client.Close()
+	})
+
+	// checker
+	checkStartJob := func(cmd string, args []string) error {
+		command := &proto.Command{
+			Cmd:  cmd,
+			Args: args,
+		}
+
+		job, err := client.Stub.StartJob(context.Background(), command)
+		if err != nil {
+			return err
+		}
+
+		logrus.Infof("Started job %s", job)
+
+		return nil
+	}
+
+	testcases := []struct {
+		name      string
+		cmd       string
+		args      []string
+		expectErr bool
+	}{
+		{
+			"Empty Command",
+			"",
+			[]string{},
+			true,
+		},
+		{
+			"Non exist command",
+			"abc",
+			[]string{},
+			true,
+		},
+		{
+			"Short term",
+			"ls",
+			[]string{"-la"},
+			false,
+		},
+		{
+			"File access",
+			"mkdir",
+			[]string{"/tmp/testdir"},
+			false,
+		},
+		{
+			"Check user",
+			"whoami",
+			[]string{},
+			false,
+		},
+		{
+			"long running",
+			"top",
+			[]string{"-b"},
+			false,
+		},
+		{
+			"High privilege",
+			"apt",
+			[]string{"update"},
+			false,
+		},
+		{
+			"sudo",
+			"sudo",
+			[]string{"apt", "update"},
+			false,
+		},
+		{
+			"bad args",
+			"ls",
+			[]string{"-wrong"},
+			false,
+		},
+	}
+
+	for _, testCase := range testcases {
+		t.Run(testCase.name, func(t *testing.T) {
+			err := checkStartJob(testCase.cmd, testCase.args)
+
+			if (err != nil) != (testCase.expectErr) {
+				t.Errorf("Test case %s with input %s returns error %v, while %v error is expected", testCase.name, testCase.cmd, err, testCase.expectErr)
+				return
+			}
+		})
+	}
+}
+
+func TestGetJobStatus(t *testing.T) {
+	server, err := service.NewServer(7777)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	go server.Serve()
+
+	client, err := client.New("127.0.0.1:7777")
+	if err != nil {
+		t.Fatalf("Fail to init client %s", err)
+	}
+
+	defer t.Cleanup(func() {
+		server.Close()
+		client.Close()
+	})
+
+	// checker
+	checkQueryJob := func(cmd string, args []string, expectedStatus *proto.ProcessStatus) error {
+		command := &proto.Command{
+			Cmd:  cmd,
+			Args: args,
+		}
+
+		job, err := client.Stub.StartJob(context.Background(), command)
+		if err != nil {
+			return err
+		}
+
+		logrus.Infof("Started job %s", job)
+
+		time.Sleep(time.Second)
```

Comment:
Just sleeping is not an ideal approach, it makes the test take longer than necessary and may not be long enough on a heavily loaded ci server

Reply:
I agree. Can you suggest me an alternative?

Reply:
The sleeping is just to wait for some short-running job to exit normally. That way I can predict the result of the query request. But I agree we could do better than just sleeping

Reply:
Good idea, thanks. I will do that